### DISCUSSION TITLE
求助：主路由无法设置静态路由到泡泡网关，请问有啥解决方案？

### DISCUSSION THREAD

**Discussion by cliffshanger** [kkkgo/PaoPaoGateWay#25] [问和答] [answered]
我采取旁路由的方式来进行搭建，除主路由的静态路由和IP外均参照下帖搭建：
https://www.right.com.cn/forum/forum.php?mod=viewthread&tid=8293054&page=1#pid18702816
PaoPaoGateWay 网络参数设置如下：
IP 192.168.2.3
Gateway 192.168.2.1
DNS：192.168.2.5（PaoPaoDNS）
节点自己搭的一个，PaoPaoGateWay 的clash.meta没有设置规则
PaoPaoDNS 网络参数
IP 192.168.2.5
其容器的CUSTOM_FORWARD设置成PaoPaoGateWay

PaoPaoGateWay+PaoPaoDNS可以正常运行起来不报错

我将局域网内的PC上的gateway和DNS直接设置成PaoPaoGateWay和PaoPaoDNS
可以正常科学上网，但是不稳定，使用一段时间后概率出现无法上网，把PaoPaoGateWay重启后就恢复了。
访问腾讯视频特别容易复现。
我目前能想到的修改
方案1是把aoPaoDNS改成递归DNS服务器，把PaoPaoGateWay当透明网关，添加节点外再设置clash规则来实现局域网的透明上网。
方案2虚拟ROS路由串联到主路由，把wan口连到主路由，设置成DHCP自动获取192.168.2.X，把lan口连接PaoPaoGateWay和PaoPaoDNS，这个虚拟lan网段用 192.168.3.X，在这个虚拟LAN里面的机器全部手动指定IP，网关和DNS。

这两种哪种方案更好？
基于我主路由无法设置静态路由的情况，还有没有更好的方法让泡泡系列完美的使用？
求大佬指点，感谢！🙏🙏

**Comment by kkkgo** [comment, marked as answer]
- 如果把电脑的网关直接设置成PaoPaoGateway的IP，那么相当于所有流量都走PaoPaoGateway处理，除非你内置了clash分流规则，不然就是相当于全局走节点代理；  
- 因为所有流量都走PaoPaoGateway处理，而腾讯视频这种应用可能会有偷带宽的P2P发起大量UDP连接等行为，会导致高负载的CPU占用和内存占用，按原本的设计来说这些国内流量也不应该让PaoPaoGateway处理，而是路由器最好是硬件路由器处理；因此你说的方案1虽然是能用但作为整个局域网透明代理并不适合；
- 如果你要弄ROS等软路由的话，你可以把主路由的DHCP功能关闭，由ROS来DHCP分配IP信息，下发DNS设置为PaoPaoDNS，网关设置保持为主路由，然后DHCP 是可以通过 option 121 下发路由表的（具体设置可以网上搜索下对应的软路由系统），这样你的网络并不会多一层nat，国内网络依然走主路由，只有被分流的流量会走PaoPaoGateway，利用DHCP完成静态路由下发。
- 最后，你的主路由是啥型号？一个《路由》器居然不支持设置路由……（不会是新出的tplink吧）其实你也可以用光猫拨号，现在的光猫性能一般都很强的，静态路由、端口映射啥的功能也肯定不缺。  

这里有个实践分享很不错，你可以参考下。https://github.com/kkkgo/PaoPaoGateWay/discussions/23

**Reply by kkkgo** [reply]
出现无法上网大概率是因为CPU负载和内存占用过高导致卡死，可以适当增加资源分配，但实际上一般不会占用这么多，原因如上所述。

**Reply by cliffshanger** [reply]
感谢大佬的回复建议，我会先去尝试下DHCP下发配置的方案。
这个无语的路由是华为路由H6，因为当初的WIFI组网选了AC+AP，当初冲着性价比和稳定性去的。结果发现这货很智能化（傻瓜），很多设置都给去掉了，譬如设置静态路由

**Reply by kkkgo** [reply]
你也可以看看光猫能不能当路由器用，然后把你的华为H6直接桥接进去，可以参考：
https://www.bilibili.com/read/cv18648339/

**Reply by Shallowhave** [reply]
佬，我在ikuai 中 使用 option 121 下发 fakeip 的静态路由，发现 在linux 下不在下发默认的静态路由表

**Comment by hsj1992** [comment]
1.如果你想指定PaoPaoGateWay作为终端的网关IP，当作纯纯的旁路网关使用的话，
所有流量都会经过clash核心，那你必须加入Clash分流规则使得clash的rule模式可用，并且fastnode功能选no，才不会一启动就是GLOBAL模式。
2.规则部分因为作者默认使用的开源官方核心，你的规则如果有涉及到官方核心没有的部分功能，你还得考虑更换clash核心生成定制镜像使用。
3.一段时间后无法上网这个问题我也有过，不过估计跟你的不同，我是因为PaoPaoGateWay更新订阅走直连，然后我的订阅地址又不好直连，导致下载的YAML文件不完整且损坏，然后就卡循环（问题在PaoPaoGateWay的控制台日志里都看得到）。我的解决方法是另开一个docker跑subconverter，然后给subconerter docker加PaoPaoGateWay的http/https proxy，PaoPaoGateWay拉取subconverter转换过的订阅链接。

**Reply by cliffshanger** [reply]
感谢你的建议和提醒，我尝试过PaoPaoGateWay做旁路网关，clash加了规则，是可以正常使用，我一台手机去测试，还是会周期性断或者假死，重启症状就消失了，所以按照大佬的建议去尝试DHCP下发配置参数的方案再试试。

**Comment by kkkgo** [comment]
关于`DHCP option 121` （DHCP推送静态路由）：
以openwrt为例，找到设置`DHCP option`的地方。`option 6`是设置下发的DNS，`option 121`是设置下发的静态路由。  
推送7.0.0.0/8下一跳到10.10.10.3
![image](https://github.com/kkkgo/PaoPaoGateWay/assets/4530897/e7bb79d3-853a-4b4f-8ec1-ac86acb44cf0)

这相当于在客户端本地推送添加了静态路由表，并且可以实际减少一跳路由。利用这个功能，你可以在局域网使用软路由系统或者别的方式搭建一个DHCP服务器，取代主路由的DHCP服务器（再差的路由器就算静态路由不能设置，DHCP服务应该也可以关闭吧？），推送路由表。   
注意，安卓系统可能不支持该特性（[参考](https://issuetracker.google.com/issues/263721377?pli=1)）。

**Reply by kiss2u** [reply]
对于 RouterOS，格式有些差异，是掩码在前，然后才是下一跳的网址，比如这样：‘8,7.0.0.0,10.10.10.3’
不过 RouterOS 的 wiki 中有关 DHCP option 说明，DHCP 中设置下一条的 Code 为 121 （classless），33 （classfull）两种，两个同时设置时以 121 为准。而且，值为 16 进制，比如上面那条信息应该为: 0x080700000010101003，这里不需要单引号或双引号，如此可以正常下发，但貌似此类设置对华为 H6 无效，待大佬测试。

**Reply by Lovest20018** [reply]
设置dhcp option ，name：paopao code：121 Value：'15,198.18.0.0,192.168.8.253'
DHCP lease 里指定dhcp options或者dhcp options set的话，这个ip就获取不到ipv4地址了。
这是哪里的问题呢？

**Reply by kiss2u** [reply]
value的值要转换为16进制，而且你的客户端必须得兼容这个dhcp option，目前我设置下来 国内的tplink和华为h6都不能正常响应option 121，不过Linux是可以的。

**Reply by kkkgo** [reply]
路由器当瘦ap用应该不影响，不负责dhcp分配。客户端的话除了安卓基本都支持。

**Comment by fsiisx** [comment]
请问ipv6 dns如何设置呢？

**Reply by kkkgo** [reply]
如果主路由支持的话，可以设置不下发IPv6 DNS。

**Reply by fsiisx** [reply]
我的主路由必须要下发ipv6 dns，应该怎么设置呢？

我已经搭建好ppdns和ppgw了，用的你教程推荐的方式，目前正常使用中。但是我有几点疑问，烦请大佬解答下。
1. 主路由下的设备，我要是手动指定成别的dns了，是不是就不会走ppdns、ppgw了？
2. 假如我在手机上另外开了surge，并且不使用下发的dns，这时候是不是也不会走ppdns、ppgw？如果使用了下发的ppdns，ppgw和surge都有fakeip，是不是会出现混乱的情况？
3. 假如ppgw用的socks5方式，socks5是一个带分流的程序，那么是否就按照规则来决定直连还是代理了？

**Reply by kkkgo** [reply]
主路由下的设备，我要是手动指定成别的dns了，是不是就不会走ppdns、ppgw了？   
取决于你有没有加额外的静态路由。比如www.google.com的解析结果是74.125.20.103，你把74.125.20.103加静态路由到ppgw的话，那么他就能走。ppdns只是把想要分流的域名扔给ppgw解析成fakeip，没有走不走的说法。

假如我在手机上另外开了surge，并且不使用下发的dns，这时候是不是也不会走ppdns、ppgw？
dns也没有啥走不走的说法，你不用就不用。至于ppgw，你只有加了静态路由的才会走。

如果使用了下发的ppdns，ppgw和surge都有fakeip，是不是会出现混乱的情况？
手机端的代理原理基本上都是虚拟VPN的形式，本来就没有什么fakeip的说法。只是有一点，如果用ppdns，而你的节点的服务器域名恰好解析结果是境外IP（现在除了自建节点一般少见），并且你开了自动分流，那么你的surge流量可能跑到ppgw那里去，你把节点的域名加入nocn_list就可避免这个情况。
